docs/design/values.md

```diff
@@ -0,0 +1,920 @@
+    -   [Polymorphic types](#polymorphic-types)
+    -   [Interop with C++ `const &` and `const` methods.](#interop-with-c-const--and-const-methods)
+    -   [Escape hatches for value addresses in Carbon](#escape-hatches-for-value-addresses-in-carbon)
+-   [Initializing expressions](#initializing-expressions)
+    -   [Function calls and returns](#function-calls-and-returns)
+        -   [Deferred initialization from values and references](#deferred-initialization-from-values-and-references)
+        -   [Declared `returned` variable](#declared-returned-variable)
+-   [Pointers](#pointers)
+    -   [Reference types](#reference-types)
+    -   [Pointer syntax](#pointer-syntax)
+    -   [Syntax-free dereference and address-of](#syntax-free-dereference-and-address-of)
+    -   [Dereferencing customization](#dereferencing-customization)
+-   [`const`-qualified types](#const-qualified-types)
+-   [Lifetime overloading](#lifetime-overloading)
+-   [Alternatives considered](#alternatives-considered)
+-   [References](#references)
+
+<!-- tocstop -->
+
+## Values, objects, and expressions
+
+Carbon has both abstract _values_ and concrete _objects_. Carbon _values_ are
+things like `42`, `true`, and `i32` (a type value). Carbon _objects_ have
+_storage_ where values can be read and written. Storage also allows taking the
+address of an object in memory in Carbon.
+
+Both objects and values can be nested within each other. For example
+`(true, true)` is both a value and also contains two sub-values. When a
+two-tuple is stored somewhere, it is both a tuple-typed object and contains two
+subobjects.
+
+These terms are important components in the describing the semantics of Carbon
+code, but they aren't sufficient. We also need to explicitly and precisely talk
+about the Carbon _expressions_ that produce or reference values and objects.
+Categorizing the expressions themselves allows us to be more precise and
+articulate important differences not captured without looking at the expression
+itself.
+
+### Expression categories
+
+There are three expression categories in Carbon:
+
+-   [_Value expressions_](#value-expressions) produce abstract, read-only
+    _values_ that cannot be modified or have its address taken.
+-   [_Reference expressions_](#reference-expressions) refer to _objects_ with
+    _storage_ where a value may be read or written and the object's address can
+    be taken.
+    -   [_Durable reference expressions_](#durable-reference-expressions) are
+        reference expressions which cannot refer to _temporary_ storage, but
+        must refer to some storage that outlives the full expression.
+    -   [_Ephemeral reference expressions_](#ephemeral-reference-expressions)
+        are reference expressions which _can_ refer to temporary storage.
+-   [_Initializing expressions_](#initializing-expressions) which require
+    storage to be provided as an input and initializes an object in that
+    storage.
```

Comment:
"input" could give the impression that the storage is supposed to be a syntactically-visible operand. I'd suggest something like "implicit input" instead.

More broadly, I had a hard time wrapping my head around this category until I got to the discussion of function returns. I think it would help to mention that example much earlier, either into this section or by making it the very next section.

Reply:
I like directly mentioning the function return example. I tried rewording the text to avoid the "input" confusion in several ways, PTAL!